lib/puppet/parser/functions/clamps_hiera_defaults.rb

```diff
@@ -0,0 +1,50 @@
+#!/opt/puppet/bin/ruby
+require 'puppet/face'
+require 'yaml'
+
+module Puppet::Parser::Functions
+  newfunction(:clamps_hiera_defaults, :type => :rvalue, :arity => 0, :doc => <<-EOS
+  This function reads through the modules on disk via the resource_type face
+  It then determines the parameters for these classes and their defaults.
+  It creates hiera data binding compatible key names i.e. 'apache::user'
+  and the evaluated ruby object as the value. This preserves boolean/hash/array 
+  EOS
+) do |args|
+   
+  #Puppet.parse_config
+  environment = Puppet[:environment] || 'production'
+  resources = Puppet::Face[:resource_type, '0.0.1'].search('.*',{:extra => { 'environment' => environment }}) 
+   
+  response = Hash.new
+    resources.each do |resource|
+     resource.arguments.each do |k,v|
+        # Walk the AST types and evaluate them, Puppet 4 may need updates
+        case v.class.to_s
+          when "Puppet::Parser::AST::String"
+            result = v.evaluate('subscope')
+          when "Puppet::Parser::AST::Undef"
+            # "We don't want undef values..."
+            next 
+          #when "Puppet::Parser::AST::Variable"
+          #  # Remove the $ in the name and wrap in hiera
+          #  # inpolation to defer the lookup until runtime
+          #  result = "%{#{v.to_s[1..-1]}}"
```

Comment:
Aw, man.  This is one of those things where I don't know how to get this stuff up and running in such a way that I could put this chunk of code under a :microscope: (debugger, probably). I'm guessing there might be a place to define a `#to_yaml` method strategically which eyaml then calls and all is :green_heart:.  

I'm guessing you know how to get things set up and running. If you'd want to fire up ScreenHero with me at some point I could pair with you on just this part.  I'm chock full of meetings today, but I'm game whenever our schedules would line up.